File: python/fastapi/order-service/api/orders/domain/order_aggregate/order.py
from __future__ import annotations

from decimal import Decimal
from typing import Tuple

import pydantic
from pydantic import UUID4


class Order(pydantic.BaseModel):
    class Config:
        frozen = True
        arbitrary_types_allowed = True

    order_id: UUID4
    customer_id: UUID4
    items: Tuple[ProductOrderItem, ...]  # tuple for immutability

    @property
    def total(self) -> Decimal:
        return sum([item.total for item in self.items])


class ProductOrderItem:
    class Config:
        frozen = True

    order_id: UUID4
    product_id: UUID4
    quantity: Decimal
    unit_price: Decimal

    @property
    def total(self) -> Decimal:
        return self.quantity * self.unit_price


Order.update_forward_refs()

File: rust/actix/order-service/src/domain/order_aggregate/order.rs
use rust_decimal::Decimal;
use rust_decimal_macros::dec;
use uuid::Uuid;

pub struct Order {
    pub order_id: Uuid,
    pub customer_id: Uuid,
    pub items: Vec<ProductOrderItem>,
}

impl Order {
    pub fn total(&self) -> Decimal {
        self.items.iter().map(|x| x.total()).sum()
    }
}

pub struct ProductOrderItem {
    pub order_id: Uuid,
    pub product_id: Uuid,
    pub quantity: Decimal,
    pub unit_price: Decimal,
}

impl ProductOrderItem {
    fn order_id(&self) -> Uuid {
        self.order_id
    }

    fn total(&self) -> Decimal {
        self.quantity * self.unit_price
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_total() {
        let product_order_item_0 = ProductOrderItem {
            order_id: Uuid::new_v4(),
            product_id: Uuid::new_v4(),
            quantity: dec!(1),
            unit_price: dec!(2.00),
        };

        let order = Order {
            order_id: Uuid::new_v4(),
            customer_id: Uuid::new_v4(),
            items: vec![product_order_item_0],
        };

        assert_eq!(order.total(), dec!(2))
    }
}
